src/main/java/com/pinterest/secor/message/Message.java

```diff
@@ -44,13 +48,17 @@ public String toString() {
         return "Message{" + fieldsToString() + '}';
     }
 
-    public Message(String topic, int kafkaPartition, long offset, byte[] payload) {
+    public Message(String topic, int kafkaPartition, long offset, byte[] key, byte[] payload) {
         mTopic = topic;
         mKafkaPartition = kafkaPartition;
         mOffset = offset;
+        mKey = key;
+        if (mKey == null) {
+            mKey = EMPTY_BYTES;
```

Comment:
Keys are optional in kafka.  They can be null if the user wants a random distribution across partitions.  Of course, log compaction and tombstones makes no sense without keys.